numpy/core/src/umath/extobj.c

```diff
@@ -0,0 +1,317 @@
+#define _UMATHMODULE
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include <Python.h>
+
+#include "npy_config.h"
+
+#define PY_ARRAY_UNIQUE_SYMBOL _npy_umathmodule_ARRAY_API
+#define NO_IMPORT_ARRAY
+
+#include "npy_pycompat.h"
+
+#include "extobj.h"
+#include "numpy/ufuncobject.h"
+
+#include "ufunc_object.h"  /* for npy_um_str_pyvals_name */
+
+#if USE_USE_DEFAULTS==1
+static int PyUFunc_NUM_NODEFAULTS = 0;
+
+/*
+ * This is a strategy to buy a little speed up and avoid the dictionary
+ * look-up in the default case.  It should work in the presence of
+ * threads.  If it is deemed too complicated or it doesn't actually work
+ * it could be taken out.
+ */
+NPY_NO_EXPORT int
+ufunc_update_use_defaults(void)
+{
+    PyObject *errobj = NULL;
+    int errmask, bufsize;
+    int res;
+
+    PyUFunc_NUM_NODEFAULTS += 1;
+    res = PyUFunc_GetPyValues("test", &bufsize, &errmask, &errobj);
+    PyUFunc_NUM_NODEFAULTS -= 1;
+    if (res < 0) {
+        Py_XDECREF(errobj);
+        return -1;
+    }
+    if ((errmask != UFUNC_ERR_DEFAULT) || (bufsize != NPY_BUFSIZE)
+            || (PyTuple_GET_ITEM(errobj, 1) != Py_None)) {
+        PyUFunc_NUM_NODEFAULTS += 1;
+    }
+    else if (PyUFunc_NUM_NODEFAULTS > 0) {
+        PyUFunc_NUM_NODEFAULTS -= 1;
+    }
+    Py_XDECREF(errobj);
+    return 0;
+}
+#endif
+
+/*
+ * fpstatus is the ufunc_formatted hardware status
+ * errmask is the handling mask specified by the user.
+ * errobj is a Python object with (string, callable object or None)
+ * or NULL
+ */
+
+/*
+ * 2. for each of the flags
+ * determine whether to ignore, warn, raise error, or call Python function.
+ * If ignore, do nothing
+ * If warn, print a warning and continue
+ * If raise return an error
+ * If call, call a user-defined function with string
+ */
+
+NPY_NO_EXPORT int
+_error_handler(int method, PyObject *errobj, char *errtype, int retstatus, int *first)
+{
+    PyObject *pyfunc, *ret, *args;
+    char *name = PyBytes_AS_STRING(PyTuple_GET_ITEM(errobj,0));
+    char msg[100];
+
+    NPY_ALLOW_C_API_DEF
+
+    /* don't need C API for a simple print */
+    if (method == UFUNC_ERR_PRINT) {
+        if (*first) {
+            fprintf(stderr, "Warning: %s encountered in %s\n", errtype, name);
+            *first = 0;
+        }
+        return 0;
+    }
+
+    NPY_ALLOW_C_API;
```

Comment:
Does this buy anything? A small speedup in errorhandling, which should be rare, doesn't seem worth it.